### PR TITLE
Sync MTU value to NAD

### PR DESCRIPTION
**Pull request by yaocw2020**
**Problem:** Refer to [issue 4353](https://github.com/harvester/harvester/issues/4353)

**Solution:** 
- Sync the MTU of the first vlanconfig to the new network-attachment-definition in the nad webhook mutator.
- All the vlanconfigs under the same cluster network must have the same MTU in the vlanconfig webhook validator.

**Related Issue:** https://github.com/harvester/harvester/issues/4353

**Test plan:**

- Create a new cluster network.
- Create a new vlanconfig. It's allowed to update the MTU if there is only one vlanconfig.
- Create another vlanconfig with the MTU with the first one. It will be successful.
- Create one more vlanconfig with the different MTU with the first one. It will fail.
- Modify the MTU of one of vlanconfigs. It will fail.
- Create a new VM network under the cluster network. Check the YAML. You will find it has the same MTU config with the vlanconfigs in the filed `spec.config`.
- Delete one vlanconfig and modify the MTU of the remind vlanconfig. The MTU will not be synchronized to the VM network.

**Additional context:**
Considering the time limit, it's not a thorough solution, and we will do more things on the issue https://github.com/harvester/harvester/issues/4355.

### CODE REVIEW

File: pkg/webhook/nad/mutator.go [hunk 1]
```diff
@@ -164,3 +185,32 @@ func tagRouteOutdated(nad *cniv1.NetworkAttachmentDefinition, oldConf, newConf *
...
+func (m *Mutator) patchMTU(conf *utils.NetConf) (types.Patch, error) {
+	clusterNetwork := conf.BrName[:len(conf.BrName)-len(iface.BridgeSuffix)]
+	vcs, err := m.vcCache.List(k8slabels.Set(map[string]string{
+		utils.KeyClusterNetworkLabel: clusterNetwork,
+	}).AsSelector())
+	if err != nil {
+		return nil, err
+	}
+
+	if len(vcs) == 0 && (vcs[0].Spec.Uplink.LinkAttrs.MTU == 0 || vcs[0].Spec.Uplink.LinkAttrs.MTU == conf.MTU) {
```
Comment (futuretea) [inline pkg/webhook/nad/mutator.go:198]: if len(vcs) == 0, vcs[0] will panic ?

Comment (yaocw2020) [reply pkg/webhook/nad/mutator.go]: Updated.

**Review by starbops — COMMENTED** [review]
The VM network test plan seems good. But I tried to apply the storage network setting with a newly created ClusterNetwork & VlanConfig, the Longhorn-related Pods were all stuck at the `ContainerCreating` status with the following events:

Screen Shot 2023-08-02 at 17 35 13

By observing the NAD CR, it seems there is a conflict between the network-controller's IPAM struct and storage-network-controller's:

```
apiVersion: k8s.cni.cncf.io/v1
kind: NetworkAttachmentDefinition
metadata:
  annotations:
    storage-network.settings.harvesterhci.io: "true"
  creationTimestamp: "2023-08-02T09:16:13Z"
  finalizers:
  - wrangler.cattle.io/harvester-network-nad-controller
  - wrangler.cattle.io/harvester-network-manager-nad-controller
  generateName: storagenetwork-
  generation: 1
  labels:
    network.harvesterhci.io/clusternetwork: provider
    network.harvesterhci.io/ready: "true"
    network.harvesterhci.io/type: L2VlanNetwork
    network.harvesterhci.io/vlan-id: "1"
  name: storagenetwork-lkz92
  namespace: harvester-system
  resourceVersion: "802830"
  uid: b21cfc40-d543-4544-944c-4773a0c110d3
spec:
  config: '{"cniVersion":"0.3.1","type":"bridge","ipam":{"type":"whereabouts"},"dns":{},"bridge":"provider-br","isGateway":false,"isDefaultGateway":false,"forceAddress":false,"ipMasq":false,"mtu":9000,"hairpinMod
e":false,"promiscMode":true,"vlan":1}'
```

For reference, the VlanConfig CR is as follow:

```
apiVersion: network.harvesterhci.io/v1beta1
kind: VlanConfig
metadata:
  annotations:
    network.harvesterhci.io/matched-nodes: '["node-0","node-1"]'
  creationTimestamp: "2023-08-02T08:48:36Z"
  finalizers:
  - wrangler.cattle.io/harvester-network-vlanconfig-controller
  generation: 1
  labels:
    network.harvesterhci.io/clusternetwork: provider
  name: enp2s0
  resourceVersion: "765954"
  uid: cfcbeadc-de15-4df2-9746-5ced054d8b6a
spec:
  clusterNetwork: provider
  uplink:
    bondOptions:
      miimon: -1
      mode: active-backup
    linkAttributes:
      mtu: 9000
      txQLen: -1
    nics:
    - enp2s0
```

**Comment by guangbochen** [comment]
@starbops seems the MTU is not added in the NAD config, did u patched the network-controller-webhook image?

**Comment by yaocw2020** [comment]
The normal config is like below
```
'{"cniVersion":"0.3.1","type":"bridge","dns":{},"bridge":"test-br","mtu":9000,"promiscMode":true,"vlan":1,"ipam":{"type":"whereabouts","range":"192.168.70.0/24"}}'
```

The `ipam.range` disappears because of the JSON marshaling.

**Comment by starbops** [comment]
> @starbops seems the MTU is not added in the NAD config, did u patched the network-controller-webhook image?

Yes, I have updated the network-controller-webhook with the PR included. There is an MTU key in `.spec.config`, which sets to 9000. Or did you mean there should be another MTU field?

**Comment by guangbochen** [comment]
aha, NVM I didn't scroll the text to the right side, P

**Review by futuretea — APPROVED** [review]
LGTM, thanks.test plan passed

**Review by guangbochen — APPROVED** [review]
LGTM, the MTU is configured correctly 
```
ip a | grep vlan
3: eno50: <BROADCAST,MULTICAST,SLAVE,UP,LOWER_UP> mtu 9000 qdisc mq master vlan-bo state UP group default qlen 1000
100: vlan-br: <BROADCAST,MULTICAST,PROMISC,UP,LOWER_UP> mtu 9000 qdisc noqueue state UP group default
122: vlan-bo: <BROADCAST,MULTICAST,MASTER,UP,LOWER_UP> mtu 9000 qdisc noqueue master vlan-br state UP group default qlen 1000
124: veth8358d68f@if4: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc noqueue master vlan-br state UP group default
127: vethd5572f17@if4: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc noqueue master vlan-br state UP group default
128: vethf10e365d@if4: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc noqueue master vlan-br state UP group default
```
also prefer to add some unit testing e.g., refer to [here](https://github.com/harvester/harvester/blob/master/pkg/webhook/resources/setting/validator_test.go).

**Comment by yaocw2020** [comment]
> LGTM, the MTU is configured correctly
> 
> ```
> ip a | grep vlan
> 3: eno50: <BROADCAST,MULTICAST,SLAVE,UP,LOWER_UP> mtu 9000 qdisc mq master vlan-bo state UP group default qlen 1000
> 100: vlan-br: <BROADCAST,MULTICAST,PROMISC,UP,LOWER_UP> mtu 9000 qdisc noqueue state UP group default
> 122: vlan-bo: <BROADCAST,MULTICAST,MASTER,UP,LOWER_UP> mtu 9000 qdisc noqueue master vlan-br state UP group default qlen 1000
> 124: veth8358d68f@if4: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc noqueue master vlan-br state UP group default
> 127: vethd5572f17@if4: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc noqueue master vlan-br state UP group default
> 128: vethf10e365d@if4: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc noqueue master vlan-br state UP group default
> ```
> 
> also prefer to add some unit testing e.g., refer to [here](https://github.com/harvester/harvester/blob/master/pkg/webhook/resources/setting/validator_test.go).

We prefer to add the unit tests for the enhancement issue https://github.com/harvester/harvester/issues/4355.

**Review by starbops — APPROVED** [review]
Test plan passed and the storage network issue is fixed. LGTM, thanks!

**Pull request merged**